### PR TITLE
[receiver/jmxreceiver] add temp_folder to jmx receiver config

### LINKED ISSUE #14757: jmxreceiver won't start in otel/opentelemetry-collector-contrib:0.61.0 docker image
## Description
The collector fails trying to start the jmxreceiver because it can't create the `jxm-config-*.properties` file.

## Steps to Reproduce
You can just apply the yaml configuration in a k8s cluster and the agent container won't start.

### Collector version

otel/opentelemetry-collector-contrib:0.61.0 docker image

### Environment information

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: collector-config
data:
  collector.yaml: |
    receivers:
      otlp:
        protocols: 
          grpc:
    processors:
    exporters:
      logging:
      prometheus:
        endpoint: "0.0.0.0:4318"
        namespace: test-space
        send_timestamps: true
        metric_expiration: 180m
        enable_open_metrics: true
        resource_to_telemetry_conversion:
          enabled: true
    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: []
          exporters: [logging]
        metrics:
          receivers: [otlp]
          processors: []
          exporters: [logging]
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: agent-config
data:
  agent.yaml: |
    receivers:
      otlp:
        protocols: 
          grpc:
      prometheus:
        config:
          scrape_configs:
            - job_name: 'otel-collector'
              scrape_interval: 5s
              metrics_path: /stats/prometheus
              static_configs:
                - targets: ['0.0.0.0:9901']
      jmx:
        jar_path: /data/otel-jmx.jar
        endpoint: localhost:1101
        target_system: jvm
        collection_interval: 10s
        # optional: the same as specifying OTLP receiver endpoint.
        #otlp:
        #  endpoint: mycollectorotlpreceiver:4317
        # username: my_jmx_username
        # determined by the environment variable value
        # password: $MY_JMX_PASSWORD
        #resource_attributes: my.attr=my.value,my.other.attr=my.other.value
        log_level: info
    processors:
      k8sattributes:
        auth_type: "serviceAccount"
        passthrough: false
        filter:
          node_from_env_var: KUBE_NODE_NAME
        extract:
          metadata:
            - k8s.pod.name
            - k8s.pod.uid
            - k8s.deployment.name
            - k8s.cluster.name
            - k8s.namespace.name
            - k8s.node.name
            - k8s.pod.start_time
        pod_association:
          - from: resource_attribute
            name: k8s.pod.uid     
      metricstransform:
        transforms:
          - include: ^envoy\_
            match_type: regexp
            action: update
            operations:
              - action: add_label
                new_label: pod_name
                new_value: ${KUBE_POD_NAME}
              - action: add_label
                new_label: test_label
                new_value: hardcoded_value
          - include: envoy_vhost_vcluster_upstream_rq_retry_overflow
            action: update
            operations:
              - action: add_label
                new_label: test_label_two
                new_value: hardcoded_value
            
    exporters:
      otlp:
        endpoint: "opentelemetrycollector.default.svc.cluster.local:4317"
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: []
          exporters: [otlp]
        metrics:
          receivers: [jmx, prometheus]
          processors: [metricstransform]
          exporters: [otlp]
---
apiVersion: v1
kind: Service
metadata:
  name: opentelemetrycollector
spec:
  ports:
    - name: grpc-otlp
      port: 4317
      protocol: TCP
      targetPort: 4317
    - name: prometheus
      port: 4318
      protocol: TCP
      targetPort: 4318
  selector:
    app.kubernetes.io/name: opentelemetrycollector
  type: ClusterIP
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: opentelemetrycollector
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: opentelemetrycollector
  template:
    metadata:
      labels:
        app.kubernetes.io/name: opentelemetrycollector
    spec:
      containers:
        - name: otelcol
          args:
            - --config=/conf/collector.yaml
          image: otel/opentelemetry-collector:0.61.0
          volumeMounts:
            - mountPath: /conf
              name: collector-config
      volumes:
        - configMap:
            items:
              - key: collector.yaml
                path: collector.yaml
            name: collector-config
          name: collector-config
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gabi
  labels:
    app: gabi
spec:
  replicas: 2
  selector:
    matchLabels:
      app: gabi
  template:
    metadata:
      labels:
        app: gabi
    spec:
      serviceAccountName: otel-collector-service-account
      initContainers:
        - name: config-data
          image: ubuntu:xenial
          command: [ "/bin/sh","-c" ]
          args: [ "apt update; apt install -y wget tar; wget -O /data/otel-jmx.jar https://repo1.maven.org/maven2/io/opentelemetry/contrib/opentelemetry-jmx-metrics/1.15.0-alpha/opentelemetry-jmx-metrics-1.15.0-alpha.jar" ]
          volumeMounts:
            - mountPath: /data
              name: jmx-jar
      containers:
        - name: app
          image: clojure:temurin-17-lein-2.9.10-bullseye
          command: ["/bin/sh", "-c", "apt-get update && apt-get install -y git && git clone https://github.com/spring-guides/gs-spring-boot.git && cd gs-spring-boot/complete/ && ./gradlew bootRun"]
          env:
            - name: JVM_OPTS
              value: "-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=1101 -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false"
        - name: envoy
          image: envoyproxy/envoy-dev:d00e4536195ec03b305a7411409910c7101aea25
          command: ["/bin/sh", "-c", "apt-get update && apt-get install -y wget && wget -O /tmp/envoy.yaml https://gist.githubusercontent.com/ggiussi/406229725fbd929a353d047fa4a6b513/raw/92462ab816145f5926c3765f6f7923915c53694d/envoy.yaml && envoy -c /tmp/envoy.yaml"]
          securityContext:
            allowPrivilegeEscalation: false
            runAsUser: 0
        - name: agent
          image: otel/opentelemetry-collector-contrib:0.61.0
          args:
          - --config=/conf/agent.yaml
          volumeMounts:
          - mountPath: /conf
            name: agent-config
          - mountPath: /data
            name: jmx-jar
          env:
            - name: KUBE_NODE_NAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: spec.nodeName
            - name: KUBE_POD_IP
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: status.podIP
            - name: KUBE_POD_NAME
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.name
        - name: sample-app
          image: springcommunity/spring-framework-petclinic:latest
          ports:
            - containerPort: 8080
      volumes:
        - name: agent-config
          configMap:
            items:
            - key: agent.yaml
              path: agent.yaml
            name: agent-config
        - name: jmx-jar
          hostPath:
            path: /mnt/kube-data/sdc-ipxe/
            type: DirectoryOrCreate
```

### OpenTelemetry Collector configuration

```yaml
receivers:
      otlp:
        protocols: 
          grpc:
      prometheus:
        config:
          scrape_configs:
            - job_name: 'otel-collector'
              scrape_interval: 5s
              metrics_path: /stats/prometheus
              static_configs:
                - targets: ['0.0.0.0:9901']
      jmx:
        jar_path: /data/otel-jmx.jar
        endpoint: localhost:1101
        target_system: jvm
        collection_interval: 10s
        # optional: the same as specifying OTLP receiver endpoint.
        #otlp:
        #  endpoint: mycollectorotlpreceiver:4317
        # username: my_jmx_username
        # determined by the environment variable value
        # password: $MY_JMX_PASSWORD
        #resource_attributes: my.attr=my.value,my.other.attr=my.other.value
        log_level: info
    processors:
      k8sattributes:
        auth_type: "serviceAccount"
        passthrough: false
        filter:
          node_from_env_var: KUBE_NODE_NAME
        extract:
          metadata:
            - k8s.pod.name
            - k8s.pod.uid
            - k8s.deployment.name
            - k8s.cluster.name
            - k8s.namespace.name
            - k8s.node.name
            - k8s.pod.start_time
        pod_association:
          - from: resource_attribute
            name: k8s.pod.uid     
      metricstransform:
        transforms:
          - include: ^envoy\_
            match_type: regexp
            action: update
            operations:
              - action: add_label
                new_label: pod_name
                new_value: ${KUBE_POD_NAME}
              - action: add_label
                new_label: test_label
                new_value: hardcoded_value
          - include: envoy_vhost_vcluster_upstream_rq_retry_overflow
            action: update
            operations:
              - action: add_label
                new_label: test_label_two
                new_value: hardcoded_value
            
    exporters:
      otlp:
        endpoint: "opentelemetrycollector.default.svc.cluster.local:4317"
        tls:
          insecure: true
    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: []
          exporters: [otlp]
        metrics:
          receivers: [jmx, prometheus]
          processors: [metricstransform]
          exporters: [otlp]
```

### Log output

```shell
2022-10-06T17:09:10.055Z        info    pipelines/pipelines.go:102      Receiver is starting... {"kind": "receiver", "name": "jmx", "pipeline": "metrics"}
Error: cannot start pipelines: failed to get tmp file for jmxreceiver config: open /tmp/jmx-config-4078735325.properties: no such file or directory
2022-10-06 17:09:10.055686 I | collector server run finished with error: cannot start pipelines: failed to get tmp file for jmxreceiver config: open /tmp/jmx-config-4078735325.properties: no such file or directory
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Adds a configuration property allowing the user to decide where to place the temporary properties file used by the JMX receiver Java process.

**Link to tracking Issue:** 
Fixes #14757

**Testing:**
Unit tests

**Documentation:**
Added to README

### CODE REVIEW

File: receiver/jmxreceiver/config.go [hunk 1]
```diff
@@ -71,6 +71,8 @@ type Config struct {
 	// Log level used by the JMX metric gatherer. Should be one of:
 	// `"trace"`, `"debug"`, `"info"`, `"warn"`, `"error"`, `"off"`
 	LogLevel string `mapstructure:"log_level"`
+	// The temporary folder used to store configuration for the JVM (uses the OS temporary folder by default).
+	TempFolder string `mapstructure:"temp_folder"`
```
Comment (rmfitzpatrick) [inline receiver/jmxreceiver/config.go:75]: 😅 

```suggestion
	TempDir string `mapstructure:"tmp_dir"`
```

Comment (atoulme) [reply receiver/jmxreceiver/config.go:75]: :)

File: .chloggen/configure-temp-dir-jmx.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add `temp_dir` configuration key to JMX receiver
+
+# One or more tracking issues related to the change
+issues: [14757]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: Adds a configuration property allowing the user to decide where to place the temporary properties file used by the JMX receiver Java process.
```
Comment (TylerHelmuth) [inline .chloggen/configure-temp-dir-jmx.yaml:16]:
```suggestion
subtext: Adds an optional configuration property allowing the user to decide where to place the temporary properties file used by the JMX receiver Java process.
```

File: receiver/jmxreceiver/config_test.go [hunk 3]
```diff
@@ -75,6 +76,7 @@ func TestLoadConfig(t *testing.T) {
 				ResourceAttributes: map[string]string{
 					"one": "two",
 				},
+				TempDir: os.TempDir(),
```
Comment (TylerHelmuth) [inline receiver/jmxreceiver/config_test.go:79]: Please add a test that tests the `temp_dir` field can be set correctly

**Review by TylerHelmuth — COMMENTED** [review]
@atoulme it looks like the receiver's README details the different config options so we'll need an update to the README for the new option.

File: receiver/jmxreceiver/config_test.go (hunk 3 above)
Comment (atoulme) [reply receiver/jmxreceiver/config_test.go:79]: test is here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17884/files#diff-11733262599e85e2533ea74b44796f627ba5c79b13c9e14599d4c4ac9baf44ccR43

**Comment by atoulme** [comment]
@TylerHelmuth added to README, thanks!

File: receiver/jmxreceiver/config_test.go (hunk 3 above)
Comment (TylerHelmuth) [reply receiver/jmxreceiver/config_test.go:79]: @atoulme I agree that that test covers the processors ability to use the configuration option correctly, and therefore also tests that the Go field can be set.  What it doesn't test is the ability for a yaml config file to be read in and then correctly map the yaml field to the Go struct's field.  For that we normally use `config_test.go` and the yaml files in `testdata`.

Comment (atoulme) [reply receiver/jmxreceiver/config_test.go]: Good point! https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17884/commits/4cfc8aec40effe991cfb46c307445d7f3c617286 covers this. Thanks!

File: receiver/jmxreceiver/README.md [hunk 4]
```diff
@@ -182,5 +184,9 @@ SLF4J log level for the JMX metrics gatherer. Must be one of: `"trace"`, `"debug
 
 Corresponds to the `org.slf4j.simpleLogger.defaultLogLevel` property.
 
+### temp_dir (default: operating system temporary directory)
+
+The location of the directory where the receiver will store configuration used by the Java process. The configuration is typically stored as a properties file, as a temporary file that is deleted on exit of the collector.
```
Comment (TylerHelmuth) [inline receiver/jmxreceiver/README.md:189]:
```suggestion
The location of the directory where the receiver will store configuration used by the Java process. The configuration is typically stored as a properties file, which is deleted on exit of the collector.
```

**Review by TylerHelmuth — APPROVED** [review]

**Review by djaglowski — CHANGES_REQUESTED** [review]
I believe this is contrary to a security related design decision proposed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6750#issuecomment-1102964088). This receiver was considered a notable security risk and the properties file was locked down as part of mitigating the risk.

The idea behind the changes was to minimize a malicious actor's ability to control the actual command that runs the subprocess. The name and path of the properties file is a part of that command, therefore it was decided that it should not be directly controlled by the user.

This is not to say that no solution like this is allowable, but I'm blocking this until at least this consideration is discussed and we have broader consensus on such a change.

**Comment by TylerHelmuth** [comment]
😮 thanks for the background @djaglowski

**Comment by rmfitzpatrick** [comment]
The subprocess functionality is provided by `os/exec` which forks using the [hardcoded java executable path](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/jmxreceiver/receiver.go#L95) so escaping shouldn't be possible from user input alone: https://github.com/golang/go/blob/f86f9a3038eb6db513a0ea36bc2af7a13b005e99/src/os/exec_posix.go#L54. Granted this isn't immune to malicious activity but if someone can arbitrarily modify the collector executable then the workflow is already compromised.

**Comment by djaglowski** [comment]
> if someone can arbitrarily modify the collector executable then the workflow is already compromised.

I don't believe this is the concern. The attack vector would be to alter the config in such a way that it effectively injects arbitrary arguments.

**Comment by rmfitzpatrick** [comment]
> > if someone can arbitrarily modify the collector executable then the workflow is already compromised.
> 
> I don't believe this is the concern. The attack vector would be to alter the config in such a way that it effectively injects arbitrary arguments.

@djaglowski do you mind explaining what you mean by injecting arbitrary arguments?

fwiw https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/os/tempfile.go;l=30 suggests that new files are always created by `os.CreateTemp()` or it will exhaustively fail.

Considering the metric gatherer can read properties from stdin it may be better to avoid the tmp file altogether.

**Comment by atoulme** [comment]
> Considering the metric gatherer can read properties from stdin it may be better to avoid the tmp file altogether.
Sounds ideal to me. There is no rush for this enhancement, so I will leave this open for comment for a couple days, afterwards if no other resolution is offered, I will close this PR and open a new ticket to move that we remove the need for the file altogether.

**Comment by djaglowski** [comment]
@rmfitzpatrick, what I mean is maliciously defined properties in the properties file. 

If we were to allow the properties file to be configurable, then a malicious actor would need very few permissions to:
1. Write a properties file with maliciously defined properties
2. Alter the config file to reference the properties file

**Comment by atoulme** [comment]
I am coming back to this as the temporary directory as understood by golang is also read from an environment variable:

https://pkg.go.dev/os#TempDir

**Comment by atoulme** [comment]
I played a bit with different values of TMPDIR and I guess that's going to be our guidance for now - use TMPDIR to set a different temporary folder rather than use a config setting. When trying to set TMPDIR to exotic values to try to interpolate a script execution in the command, the go program fails with messages such as these:
```
go: creating work dir: stat /tmp;touch /tmp/foo: no such file or directory
```

**Pull request closed without merging**